### PR TITLE
Implement Basic LMD Ghost Fork Choice Skeleton + Helper Funcs

### PR DESCRIPTION
**Pull request by rauljordan**
This is part of #1307.

---

# Description

This PR implements the basic skeleton needed for the Latest Message Driven fork-choice rule for Ethereum Serenity. This is based on the pseudocode specified by the official spec here: https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#beacon-chain-fork-choice-rule

This PR only implements the pure functions needed for LMD. It does not yet integrate it into the blockchain service.

- [x] Add basic LMD tests
- [x] Add tests for all LMD helper functions
- [x] Benchmark LMD: for default validator count, it takes 1 minute currently to select a new head

### CODE REVIEW

File: beacon-chain/blockchain/fork_choice.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+// beacon block's hash as their parent root hash.
+func BlockChildren(block *pb.BeaconBlock, observedBlocks []*pb.BeaconBlock) ([]*pb.BeaconBlock, error) {
+	var children []*pb.BeaconBlock
+	encoded, err := proto.Marshal(block)
+	if err != nil {
+		return nil, fmt.Errorf("could not marshal block: %v", err)
+	}
+	hash := hashutil.Hash(encoded)
+	for _, observed := range observedBlocks {
+		if bytes.Equal(observed.ParentRootHash32, hash[:]) {
```
Comment (nisdas) [inline beacon-chain/blockchain/fork_choice.go]: Could there be a better way of ordering the observed blocks ? For now its an unsorted array, but if we are looking for children we can only have observed blocks with slots more than the current slot of the head. Maybe group them by slot if possible. But these optimizations can be done in another PR

Comment (rauljordan) [reply beacon-chain/blockchain/fork_choice.go]: Yes I think these helpers will be heavily refactored in an optimization PR, just wanted to get something basic that works

File: beacon-chain/blockchain/fork_choice_test.go [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+	}
+	higherVoteBlock := &pb.BeaconBlock{
+		Slot:             5,
+		ParentRootHash32: genesisHash[:],
+		StateRootHash32:  []byte("some-other-state"),
+	}
+	higherVoteBlockEnc, _ := proto.Marshal(higherVoteBlock)
+	higherVoteBlockHash := hashutil.Hash(higherVoteBlockEnc)
+
+	// We store these potential heads in the DB.
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply beacon-chain/blockchain/fork_choice_test.go]: These tests will be refactored soon in a new PR, also I believe they are not as comprehensive and will need to model more complex scenarios as part of those optimization PRs.

File: beacon-chain/db/attestation.go [hunk 3]
```diff
@@ -43,6 +43,44 @@ func (db *BeaconDB) GetAttestation(hash [32]byte) (*pb.Attestation, error) {
...
+	return db.update(func(tx *bolt.Tx) error {
+		a := tx.Bucket(attestationBucket)
+		key := []byte(fmt.Sprintf("validator-%d-latest-att", validatorIdx))
+		return a.Put(key, encodedAtt)
+	})
+}
+
+// GetLatestAttestationForValidator returns the observed attestations by a validator at a
+// certain index in the beacon state's registry.
+func (db *BeaconDB) GetLatestAttestationForValidator(validatorIdx uint32) (*pb.Attestation, error) {
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply beacon-chain/db/attestation.go]: https://github.com/prysmaticlabs/prysm/issues/1337 will do this in a PR that addresses this issue

**Comment by rauljordan** [comment]
PTAL @prestonvanloon

File: beacon-chain/blockchain/fork_choice.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+		blockHash := hashutil.Hash(blockEnc)
+		if blockHash == ancestorHash {
+			votes++
+		}
+	}
+	return votes, nil
+}
+
+// BlockAncestor obtains the ancestor at of a block at a certain slot.
+func BlockAncestor(block *pb.BeaconBlock, slot uint64, beaconDB *db.BeaconDB) (*pb.BeaconBlock, error) {
```
Comment (terenc3t) [inline beacon-chain/blockchain/fork_choice.go]: A more appropriate place for `BlockAncestor ` is under `block.go`. `BlockAncestor` is more of a block-specific operation, not fork choice

File: beacon-chain/blockchain/fork_choice.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	}
+	if parent == nil {
+		return nil, fmt.Errorf("parent block does not exist: %v", err)
+	}
+	return BlockAncestor(parent, slot, beaconDB)
+}
+
+// BlockChildren obtains the blocks in a list of observed blocks which have the current
+// beacon block's hash as their parent root hash.
+func BlockChildren(block *pb.BeaconBlock, observedBlocks []*pb.BeaconBlock) ([]*pb.BeaconBlock, error) {
```
Comment (terenc3t) [inline beacon-chain/blockchain/fork_choice.go]: Same as `BlockAncestor`, I think it's better under `block.go`

File: beacon-chain/blockchain/fork_choice.go (hunk 4 above)
Comment (rauljordan) [reply beacon-chain/blockchain/fork_choice.go]: Ancestor is very coupled with the vote count and currently depends on the DB. I don't want to introduce db methods into core, so for now we can just move children. Also, the ancestor function is the prime candidate to be changed with future optimizations.

File: beacon-chain/blockchain/fork_choice.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+// VoteCount determines the number of votes on a beacon block by counting the number
+// of target blocks that have such beacon block as a common ancestor.
+func VoteCount(block *pb.BeaconBlock, targets map[[32]byte]*pb.BeaconBlock, beaconDB *db.BeaconDB) (int, error) {
+	votes := 0
+	for k := range targets {
+		ancestor, err := BlockAncestor(targets[k], block.Slot, beaconDB)
+		if err != nil {
+			return 0, err
+		}
+		ancestorEnc, err := proto.Marshal(ancestor)
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/fork_choice.go]: Linkwise here, use block hash method in hashutil

**Pull request merged**